arangod/Replication2/ReplicatedLog/InMemoryLog.cpp

```diff
@@ -0,0 +1,312 @@
+#include <immer/flex_vector_transient.hpp>
+#if (_MSC_VER >= 1)
+#pragma warning(pop)
+#endif
+
+using namespace arangodb;
+using namespace arangodb::replication2;
+
+auto replicated_log::InMemoryLog::getLastIndex() const noexcept -> LogIndex {
+  auto const result = LogIndex{_log.size()};
+  // log empty => result == 0
+  TRI_ASSERT(!_log.empty() || result == LogIndex(0));
+  // !log empty => result index == last entry
+  TRI_ASSERT(_log.empty() || result == _log.back().entry().logIndex());
+  return result;
+}
+
+auto replicated_log::InMemoryLog::getLastTermIndexPair() const noexcept -> TermIndexPair{
+  if (_log.empty()) {
+    return {};
+  }
+  return _log.back().entry().logTermIndexPair();
+}
+
+auto replicated_log::InMemoryLog::getLastTerm() const noexcept -> LogTerm {
+  if (_log.empty()) {
+    return LogTerm{0};
+  }
+  return _log.back().entry().logTerm();
+}
+
+auto replicated_log::InMemoryLog::getNextIndex() const noexcept -> LogIndex {
+  return getLastIndex() + 1;
+}
+
+auto replicated_log::InMemoryLog::getEntryByIndex(LogIndex const idx) const noexcept
+    -> std::optional<InMemoryLogEntry> {
+  if (_log.size() < idx.value || idx.value == 0) {
+    return std::nullopt;
+  }
+
+  auto const& e = _log.at(idx.value - 1);
+  TRI_ASSERT(e.entry().logIndex() == idx);
+  return e;
+}
+
+auto replicated_log::InMemoryLog::splice(LogIndex from, LogIndex to) const -> log_type {
+  from = LogIndex{std::max<decltype(from.value)>(from.value, 1)};
+  auto res = _log.take(to.value - 1).drop(from.value - 1);
+  TRI_ASSERT(res.size() == to.value - from.value);
+  return res;
+}
+
+auto replicated_log::InMemoryLog::getFirstIndexOfTerm(LogTerm term) const noexcept -> std::optional<LogIndex> {
+  auto it = std::lower_bound(_log.begin(), _log.end(), term,
+                             [](auto const& entry, auto const& term) {
+                               return term > entry.entry().logTerm();
+                             });
+
+  if (it != _log.end() && it->entry().logTerm() == term) {
+    return it->entry().logIndex();
+  } else {
+    return std::nullopt;
+  }
+}
+
+auto replicated_log::InMemoryLog::getLastIndexOfTerm(LogTerm term) const noexcept
+    -> std::optional<LogIndex> {
+  // Note that we're using reverse iterators
+  auto it = std::lower_bound(_log.rbegin(), _log.rend(), term,
+                             [](auto const& entry, auto const& term) {
+                               // Note that this is flipped
+                               return entry.entry().logTerm() > term;
+                             });
+
+  if (it != _log.rend() && it->entry().logTerm() == term) {
+    return it->entry().logIndex();
+  } else {
+    return std::nullopt;
+  }
+}
+
+replicated_log::InMemoryLog::InMemoryLog(LoggerContext const& logContext,
+                                         replicated_log::LogCore const& logCore) {
+  auto iter = logCore.read(LogIndex{0});
+  auto log = _log.transient();
+  while (auto entry = iter->next()) {
```

Comment:
Should we make a reservation for `log` here in order to save some reallocations? This should be possible if the `logCore` here can provide the number of its log entries.

Reply:
We already use a transient object that is much more efficient than the immutable object at creating a new state. However, according to the [docs](https://sinusoid.es/immer/transients.html#flex-vector-transient) it does not provide a reserve.